blmath/numerics/test_coercion.py

```diff
@@ -42,7 +43,11 @@ def test_as_numeric_array_three_by_one(self):
                 as_numeric_array(v, shape=(3,), allow_none=False)
 
         for v in good_iff_allow_none_and_empty_as_none:
-            as_numeric_array(v, shape=(3,), allow_none=True, empty_as_none=True)
+            with warnings.catch_warnings(record=True) as w:
+                warnings.simplefilter("always")
+                as_numeric_array(v, shape=(3,), allow_none=True, empty_as_none=True)
+                assert len(w) == 1
+                assert issubclass(w[-1].category, DeprecationWarning)
```

Comment:
This call to `as_numeric_array` throws a deprecation warning (https://github.com/lace/blmath/blob/master/blmath/numerics/coercion.py#L35). Before this, the warning was being thrown, but not checked.

This is two reasons why I've mostly switched to pytest over nose2/unittest. pytest collects the warnings thrown and shows them to you after a test run, so you can see things like deprecation warnings. There are a bunch of them that this test suite throws, for old usages of numpy mostly. I didn't bother to fix them, on account of this package being end of life. But I didn't want this one intentional deprecation warning mixed in with all of the actual deprecation warnings this code throws.

The other reason I've switched to pytest for testing is that it's so much cleaner. This test could read:

```py
with pytest.warns(DeprecationWarning):
    as_numeric_array(v, shape=(3,), allow_none=True, empty_as_none=True)
```

Reply:
Gotcha! Thanks for the explanation. I did not realize pytest did that – it's super cool!!

If we weren't heading to EOL I'd say we should add a comment in the source, but when the code under test moves, the test will be running in pytest anyway so no need.

FWIW in new code I've replaced the calls to `as_numeric_array` with [`vg.shape.check`](https://vgpy.readthedocs.io/en/latest/#vg.shape.check). I'd be curious your thoughts on that approach. One thing I've noticed is that `locals()` is a little bit slow, so it may be worth deprecating that method in favor of directly invoking [`vg.shape.check_value`](https://vgpy.readthedocs.io/en/latest/#vg.shape.check_value).

Reply:
Yeah, I like the clarity of moving from "magically make it shaped like this" to more of a precondition assertion. That said, I don't really see the value of `.check` over `.check_value` here. Embedding the call to `locals()` makes it kind of an anti-sugar and opens up the idea of passing around `locals()` in scary ways... Why would you want to 

```py
def my_fun_function(points):
    k = vg.shape.check(locals(), 'points', (-1, 3))
```

vs.

```py
def my_fun_function(points):
    k = vg.shape.check_value(points, (-1, 3))
```

Reply:
The idea of `check()` is that you can have "points" appear in the error message – which is very helpful – without needing to pass in both `points` and `name="points"`.

Reply:
If it's for code that only runs in the exception case, why not move the slow bit to only run then? We can always walk the stack to get locals:

```py
import inspect
caller_locals = inspect.currentframe().f_back.f_locals
names = "/".join([k for k, v in caller_locals.items() if v is a])
print(f"check({names})")
```

The only tricky bit is that there might be several names that refer to the same object in that namespace, or no name at all if it's an unnamed value.

Reply:
I see – and then look for a value that `is` the object that was passed in.

I definitely don't mind a bit of magic in the exception case!

Is there a way to get the code from the call to `check()` or the name of the variable that was passed in?

Reply:
I see – and then look for a value that `is` the object that was passed in.

I definitely don't mind a bit of magic in the exception case!

Is there a way to get the code from the call to `check()` or the name of the variable that was passed in?

Reply:
Sort of... we're starting to get pretty fragile if we go down that route:

```py
import inspect
caller_frame = inspect.currentframe().f_back
caller_local = caller_frame.f_locals.items()
names = "/".join([k for k, v in caller_local if v is x])
print("called with {}".format(names))
source = inspect.getsource(caller_frame.f_code)
try:
    print(source.split("\n")[caller_frame.f_lineno - caller_frame.f_code.co_firstlineno])
except:
    print("not found")
```

will get you the calling line _most_ of the time. It wouldn't work when the call is at the top level of the file, when the function call is spread across several lines, or maybe some other random times. Plus it's just going to give you the text of the call, so it might not really be what you need. Personally, I think it's too fragile...